### PR TITLE
ipc4: helper: Fix module unbinding 

### PR DESCRIPTION
**Pull request by ranj063**
When 2 modules are bound together, they are bound with a buffer between them with the connection like this : source_module->buffer->sink_module.

So, what this means is that there is only one connection going from a module to the buffer. So, calling pipeline_disconnect() from the same module to the same buffer in both directions makes no sense.

The right way to free a module would be first disconnect the module from the sink buffer, then check if the sink buffer has been disconnected from its sink module. If so, the buffer should be freed. Otherwise the buffer will be freed whenever the sink module is disconnected from it. The same thing must be repeated for the module's source buffer as well.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: src/ipc/ipc4/helper.c
Comment (RanderWang) [inline src/ipc/ipc4/helper.c:205]: each component free its sink buffer then all the buffer will be free.  If a component free both sink and source buffer then the connected component will have no sink buffer or source buffer to free

Comment (ranj063) [reply src/ipc/ipc4/helper.c:205]: @RanderWang but the code frees both buffers only if the buffer isnt connected to a source and a sink and then it would be correct.

File: src/ipc/ipc4/helper.c
Comment (RanderWang) [inline src/ipc/ipc4/helper.c:196]: I am confused about sink or source.   check buffer_c->source  or buffer_c->sink to free buffer

Comment (RanderWang) [reply src/ipc/ipc4/helper.c:196]: I meant this loop is checking each sink buffer and free it if it doesn't have source buffer

Comment (ranj063) [reply src/ipc/ipc4/helper.c:196]: no @RanderWang . This loop checks each sink buffer and frees it the sink buffer doesnt have a sink comp.

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by lyakh — COMMENTED** [review]
you write "calling pipeline_disconnect() from the same module to the same buffer in both directions makes no sense." But I don't think this is currently done. Once the buffer is disconnected from *both* directions in deleted by this patch lines 193 and 194 it will be unlinked from both lists - on its sink and source connections. So when then the other component is freed, it won't find that buffer in that its list, so it won't be "freed again." Or am I missing anything?

File: src/ipc/ipc4/helper.c
Comment (kv2019i) [inline src/ipc/ipc4/helper.c:196]: Unrelated to PR, but using "sink" to refer to buffers that are sinks, and also to "sink component", is making it hard to follow related code. But that's way beyond this PR. The logic seems correct.

File: src/ipc/ipc4/helper.c
Comment (kv2019i) [inline src/ipc/ipc4/helper.c:395]: The commit could use a few words why you do this? We get some additional debug and a bit dubious handling cache invalidation logic (in pipeline_disconnect), but what's the main motivation?

File: src/ipc/ipc4/helper.c
Comment (kv2019i) [inline src/ipc/ipc4/helper.c:204]: Is this a possible case that if two modules A and B are connected, then it would be valid to for B to remain active when A is unbound? 

This change seems ok (try to free both in A and B seems sane), but wondering can this happen.

**Review by kv2019i — COMMENTED** [review]
A few comments inline. This seems like a good change, but the commit messages leave me a bit wondering what is the main motivation. Is there a big bug lurking, or is this just cleaning up the code?

File: src/ipc/ipc4/helper.c
Comment (ranj063) [reply src/ipc/ipc4/helper.c:204]: @kv2019i this is possible like in the case of mixin/mixout and they would be on different pipelines. So it would do be freed when the UNBIND_MODULES IPC is handled.

**Comment by ranj063** [comment]
> you write "calling pipeline_disconnect() from the same module to the same buffer in both directions makes no sense." But I don't think this is currently done. Once the buffer is disconnected from _both_ directions in deleted by this patch lines 193 and 194 it will be unlinked from both lists - on its sink and source connections. So when then the other component is freed, it won't find that buffer in that its list, so it won't be "freed again." Or am I missing anything?

@lyakh please have a look at https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L193 and 
https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L194

**Comment by ranj063** [comment]
> A few comments inline. This seems like a good change, but the commit messages leave me a bit wondering what is the main motivation. Is there a big bug lurking, or is this just cleaning up the code?

@kv2019i this isnt just clean up. It does the right thing. There's clearly some incorrect logic today. I do not know if I can pin point a specific issue because of the logic but I certainly have much better stress test results with this PR than without

**Comment by lyakh** [comment]
> > you write "calling pipeline_disconnect() from the same module to the same buffer in both directions makes no sense." But I don't think this is currently done. Once the buffer is disconnected from _both_ directions in deleted by this patch lines 193 and 194 it will be unlinked from both lists - on its sink and source connections. So when then the other component is freed, it won't find that buffer in that its list, so it won't be "freed again." Or am I missing anything?
> 
> @lyakh please have a look at
> 
> https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L193
> and
> 
> https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L194

yes, sorry, still don't understand:
1. `ipc_pipeline_module_free()` is entered, it enters the https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L184 line with a certain component, let's say component A
2. the loop over bsink-linked buffers at https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L190 is entered, the first buffer is taken, call it buffer P
3. that buffer is disconnected from both its sink and its source connections, components A and B

So, when when later component B is handled, buffer P isn't linked to it any more, so it isn't found and therefore no further attempt to disconnect it is made

**Comment by ranj063** [comment]
> 3\. that buffer is disconnected from both its sink and its source connections, components A and B

@lyakh no that buffer is connected from its source only with the current logic. 

> 3\. that buffer is disconnected from both its sink and its source connections, components A and B

where is the disconnection from both sink and source components done in the current code? It only disconnects from the source component (with both directions COMP_TO_BUFFER and BUFFER_TO_COMP set).

**Review by RanderWang — APPROVED** [review]
LGTM

**Comment by lyakh** [comment]
> > 3. that buffer is disconnected from both its sink and its source connections, components A and B
> 
> @lyakh where is the disconnection from both sink and source components done in the current code? It only disconnects from the source component (with both directions COMP_TO_BUFFER and BUFFER_TO_COMP set).

@ranj063 right there https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/ipc/ipc4/helper.c#L193-L196 - those two directions do mean the sink and the source direction, otherwise how would they free the buffer in the last line there?

**Review by kv2019i — APPROVED** [review]

**Comment by ranj063** [comment]
> - those two directions do mean the sink and the source direction, otherwise how would they free the buffer in the last line there?

@lyakh but those2 lines mean disconnect the connections from comp->buffer and buffer->comp where the buffer and comp are exactly the same. This is never possible. A connection from a buffer to a comp is never 2-way

**Comment by ranj063** [comment]
let's not merge this PR yet. I would like to run more stress tests with it before merging.

**Comment by lyakh** [comment]
> > * those two directions do mean the sink and the source direction, otherwise how would they free the buffer in the last line there?
> 
> @lyakh but those2 lines mean disconnect the connections from comp->buffer and buffer->comp where the buffer and comp are exactly the same. This is never possible. A connection from a buffer to a comp is never 2-way

@ranj063 I was already about to write, that you were right, but then I looked at the `pipeline_disconnect()` implementation (again). What gets disconnected there is https://github.com/thesofproject/sof/blob/9200231e0ed3eebd8b2fdc6ede5f0de17b4252ec/src/audio/pipeline/pipeline-graph.c#L221 , i.e. for unlinking the list the component isn't used. It just takes the list within the buffer in the specified direction. Also why would we otherwise have those 2 lines there? We know, that the buffer is a sink, so the direction is always component-to-buffer, isn't it? So if you were correct the second of those two lines would simply be always a NOP and could be removed. In fact, looking at these two functions now I do think there's a mistake in `ipc_pipeline_module_free()`, I think to make cache-synchronisation correct that function has to be modified as
```
	pipeline_disconnect(sink->source, sink, PPL_CONN_DIR_COMP_TO_BUFFER);
	pipeline_disconnect(sink->sink, sink, PPL_CONN_DIR_BUFFER_TO_COMP);
```
that being pseudo-code - in fact you'd need to acquire the `sink` buffer to get its sink and source into local variables, then release it, then call this code with locally saved pointers

**Review by lyakh — CHANGES_REQUESTED** [review]
@ranj063 requested this PR not to be merged for now, and I have some unclarified questions

**Comment by ranj063** [comment]
@lyakh i've updated this PR now and I think it does the right thing. It actually fixes the IPC timeouts we see during the PIPELINE_DELETE IPC.

**Comment by ranj063** [comment]
@lyakh @kv2019i @lgirdwood this PR fixes https://github.com/thesofproject/sof/issues/6465

**Review by kv2019i — APPROVED** [review]
Refreshing my +1

**Review by lyakh — APPROVED** [review]
we don't quite understand why but this seems to fix some crashes. The alternative fix to patch 2 from this PR from #6497 didn't work for some reason

**Comment by kv2019i** [comment]
One check failed, but that was caused by device being unavailable for testing and as that was a IPC3 test device, doesn't affect this PR which is touching IPC4 code. Plus all tests pass on other IPC3 test targets, so proceeding with merge.

**Pull request merged**
